INT-4252 IntegrationFlow: allow custom bean names

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4252

To allow to provide arbitrary bean names for the intermediate components
in the `IntegrationFlow` change `ComponentsRegistration` to return
a `Map<Object, String>` instead of raw `Collection<Object>`

* Use the value from that map in the `IntegrationFlowBeanPostProcessor` as a
fallback option before walking into bean name generation
* Provide `containerId` option for the AMQP DSL components

artembilan committed:
INT-4252 IntegrationFlow: allow custom bean names

JIRA: https://jira.spring.io/browse/INT-4252

To allow to provide arbitrary bean names for the intermediate components
in the `IntegrationFlow` change `ComponentsRegistration` to return
a `Map<Object, String>` instead of raw `Collection<Object>`

* Use the value from that map in the `IntegrationFlowBeanPostProcessor` as a
fallback option before walking into bean name generation
* Provide `containerId` option for the AMQP DSL components

garyrussell commented on spring-integration-amqp/src/test/java/org/springframework/integration/amqp/dsl/AmqpTests.java:
```diff
@@ -82,11 +85,9 @@
 
-	@AfterClass
-	public static void tearDown() {
-		brokerRunning.removeTestQueues("amqpOutboundInput", "amqpReplyChannel", "asyncReplies", "defaultReplyTo",
```
We still need this (but with no `additionalQueues` args).

garyrussell commented on spring-integration-amqp/src/main/java/org/springframework/integration/amqp/dsl/AmqpInboundGatewaySpec.java:
```diff
@@ -270,10 +272,19 @@ public AmqpInboundGatewaySpec defaultRequeueRejected(boolean defaultRequeueRejec
-	public Collection<Object> getComponentsToRegister() {
-		return Collections.<Object>singleton(this.listenerContainer);
+	/**
+	 * @param containerId the bean name for internal listener container instance.
```
/* Set the bean name of the listener container if it is created internally or a non-bean
    /* object was supplied. Ignored if the container is already a bean.

garyrussell replied:
I think we need this on the Jms specs that generate a container too.

artembilan committed:
* Revert `@AfterClass` in the `AmqpTests`
* Expose `id()` for the `JmsDestinationAccessorSpec`
* Register `ListenerContainer` and `JmsTemplate`
as bean via `ComponentRegistration` in the particular JMS Java DSL components
* Fix `Jms` factory to populate the `JmsDefaultListenerContainerSpec`
for the `messageDrivenChannelAdapter()` without class specified

artembilan committed:
* Refactor AMQP DSL Inbound components to deal with the new ContainerSpec API
* Remove `containerId()` in favor of `id()` in the ContainerSpec

garyrussell commented:
Merged as ea89682368a3b7946f5378fbbfba82e692894021
